### ISSUE TITLE
[feature] Allow `GTS_HOST` and `GTS_ACCOUNT_DOMAIN` to be on different second-level domains

### ISSUE THREAD

**Issue by aaro-n** [superseriousbusiness/gotosocial#2106]
### Is your feature request related to a problem ?

gotosocial requires the domain names of GTS_HOST and GTS_ACCOUNT_DOMAIN to be the same, different domain names cannot work.

### Describe the solution you'd like.

GTS_ACCOUNT_DOMAIN can be set as the only identification method in the ActivityPub network.
GTS_HOST can set any domain name, and it is the same as GTS_ACCOUNT_DOMAIN by default.

### Describe alternatives you've considered.

NONE

### Additional context.

GTS_HOST can be set to be different from GTS_ACCOUNT_DOMAIN, which is very useful against GFW blockade. Mastodon can set LOCAL_DOMAIN and ALTERNATE_DOMAINS, and Mastodon can be different from these two domain names.

**Comment by tsmethurst** [comment]
Can you explain a bit more? Very unclear from this description what the issue/request is.

**Comment by daenney** [comment]
Right now, `GTS_HOST` and `GTS_ACCOUNT_DOMAIN` can already be different. It's how you can have a split-domain setup, with the API living on `social.example.com` and but the accounts being `user@example.com`. If that's what you're trying to achieve, that already works.

I don't think we check that `GTS_HOST` has to be a subdomain of `GTS_ACCOUNT_DOMAIN`, so even a setup like `GTS_HOTS=banana.example.org` with `GTS_ACCOUNT_DOMAIN=example.cn` should already work. I might be mistaken about that.

**Comment by aaro-n** [comment]
@daenney @tsmethurst  No, gotosocial requires the same domain name, subdomains can be different, for example: "123.com" and "me.123.com" can work, "123.com" and "me.567.xyz" can't .
# Environment variables:
  GTS_HOST = "me.123.fun"
  GTS_ACCOUNT_DOMAIN = "222.ee"
# Running log (docker):
gotosocial version: superseriousbusiness/gotosocial:latest
Log: 
```
2023/08/12 22:53:46 error executing command: invalid config: host was me.123.fun and account-domain was 222.ee, but me.123.fun is not a valid subdomain of 222.ee
   INFO Main child exited normally with code: 1
   INFO Starting clean up.
```

gotosocial version: superseriousbusiness/gotosocial:0.11.0-rc2
Log: 
```
2023/08/12 22:57:20 error executing command: invalid config: host was me.123.fun and account-domain was 222.ee, but me.123.fun is not a valid subdomain of 222.ee
   INFO Main child exited normally with code: 1
   INFO Starting clean up.
```

**Comment by daenney** [comment]
Ah right. I think that we might be able to relax. It shouldn't have to be a subdomain afaik.

**Comment by igalic** [comment]
what does the spec say?

**Comment by tsmethurst** [comment]
> what does the spec say?

Bugger all if I remember rightly: this isn't something that's in the ActivityPub spec. For this we probably oughta consider looking at the webfinger spec instead, since that's the main mechanism by which we do this account domain trick.

I think it was me that wrote the code to ensure that `host` is a valid subdomain of `account-domain`, just because I figured that was the most common use case, and it seemed a good idea to at least slightly restrict this.

I'm up for changing it if there's valid use cases for having `host` and `account-domain` on totally separate second-level domains, but let's investigate first to make sure that people can't do wacky things like set their `account-domain` to google.com or whatever, and have that actually work somehow.

**Comment by daenney** [comment]
Through webfinger folks even set account aliasses. Like you can be `@a@customdomain.tld` but have that return the profile for `@b@mastodon.social`. From that point of view I think it's fine.

With the `account-domain` set to `google.com`, we'd still reach out to `google.com/.well-known/webfinger?...` etc., so unless you also control the account domain I doubt you'd be able to do anything bad with it.

**Comment by tsmethurst** [comment]
@aaro-n I edited the issue title, hope that's OK

**Comment by igalic** [comment]
> With the `account-domain` set to `google.com`, we'd still reach out to `google.com/.well-known/webfinger?...` etc., so unless you also control the account domain I doubt you'd be able to do anything bad with it.

could you use that to impersonate another account?

**Comment by daenney** [comment]
Mmm, not really impersonate as far as I can tell. But the end result would be that if someone mentioned `@a@customdomain.tld` then `@b@mastodon.social` would actually be the recipient. So in theory you could set up all kinds of rude aliases on behalf of someone and direct abuse that way. The webfinger protocol doesn't care and afaik there's no way to negotiate that.

**Comment by igalic** [comment]
in that case, I'd rather we keep the current restriction.

**Comment by aaro-n** [comment]
@igalic @daenney @tsmethurst @technomancy  Combined with my own actual usage, what I need is the mirror website of  GoToSocial , which can use or forward the API of  GoToSocial .
## Idea:
* `GTS_HOST` <--> GoToSocial Original Site（docker），Use a short domain name to make the username appear shorter and to use the domain name to build other things.
* `GTS_ACCOUNT_DOMAIN` <--> GoToSocial Original Site（docker），The main site of GoToSocial, the domain name can be of any length, it can be a third-level domain name under the GTS_HOST domain name, and it is responsible for communicating with other ActivityPub sites.
* `website mirror`  <--> GoToSocial Original Site（docker）or GTS_ACCOUNT_DOMAIN，GoToSocial mirror site, you can use GoToSocial API to log in to the client and manage the website. To communicate with other ActivityPub sites, you must use the GTS_ACCOUNT_DOMAIN site. The mirror site is only used for website blocking or website access acceleration. The mirror site does not participate in ActivityPub communication.
# Why do I need a GoToSocial mirror site?
I am from mainland China. Because of the GFW blockade and the per capita bandwidth from the mainland to overseas is very small, direct access to overseas websites is either very slow or inaccessible. I need to use CDN acceleration in the mainland. At the same time, because of government requirements, CDN operators require domain names to be filed and Only specific first-level domain names can be used. At the same time, when the website is blocked by GFW, you need to use VPN to access or replace the unblocked domain name. If you can build a website mirror, you can just buy a domain name and reverse it on the Hong Kong vps at that time, without changing the main domain name.
# Use nginx to reverse GTS_ACCOUNT_DOMAIN test
Introduction: GoToSocial is located behind cloudflare, I directly use Nginx to reverse generation, GTS_ACCOUNT_DOMAIN domain name is:  me.xxx.fun, nginx domain name is: ce.aaa.cn.
nginx configuration：
```
server {
    listen 443 ssl http2;
    listen [::]:443;

    server_name ce.aaa.cn;

    client_max_body_size 100m;
    client_body_timeout 60;

    include /home/www/nginx/config/ssl.conf;

    include /home/www/nginx/config/cdn.conf;

    include /home/www/nginx/config/gzip.conf;


    location / {
       proxy_pass https://me.xxx.fun.cdn.cloudflare.net;
       proxy_set_header Host me.xxx.fun;

       proxy_set_header Upgrade $http_upgrade;
       proxy_set_header Connection "upgrade";
       proxy_set_header X-Forwarded-For $remote_addr;
       proxy_set_header X-Forwarded-Proto $scheme;
   
       proxy_ssl_name me.xxx.fun;
       proxy_ssl_server_name on;

     }
 }
```
Actual effect: Visit ce.aaa.cn, the website can be opened, but some pictures cannot be displayed, the picture domain name address is: me.xxx.fun, when visiting ce.aaa.cn/settings, the `Instance` is https://ce .aaa.cn, after logging in, it prompts:
```
Instance Logo
我的自留地
An error occured:
Bad Request: field response_type was not set on OAuthAuthorize form: If you arrived at this error during a login/oauth flow, please try clearing your session cookies and logging in again; if problems persist, make sure you're using the correct credentials
Request ID: 3bywhwc904001q4hj2zg
```

**Comment by eallion** [comment]
I also tried to set it up like this, but it is not supported. `GTS_HOST` must be a subdomain of `GTS_ACCOUNT_DOMAIN`. They must be under the same root domain.

**Comment by aaro-n** [comment]
When GTS_HOST is different from GTS_ACCOUNT_DOMAIN, some logs of the database are filled for the first time:
# Environment variables:
```
  GTS_HOST = "me.889988.ee"
  GTS_ACCOUNT_DOMAIN = "889988.ee"
```
# LOG：
```
timestamp="19/08/2023 15:24:39.022" func=bundb.(*adminDB).CreateInstanceAccount level=INFO msg="instance account me.889988.ee CREATED with id 01RC98153YT02ES8483SPP46AJ"
imestamp="19/08/2023 15:24:39.039" func=bundb.(*adminDB).CreateInstanceInstance level=INFO msg="created instance instance me.889988.ee with id 01Q5NRBB1ZQHHGNAFHK7TY2HYY"
```

**Comment by daenney** [comment]
We've migrated to Codeberg. Your issue has been imported as https://codeberg.org/superseriousbusiness/gotosocial/issues/2106.
